File: go/0077M-组合.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-12 15:00:56
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

func combine(n int, k int) [][]int {
	ans := make([][]int, 0)
	vis := make([]int, 0)
	var dfs func(p int)
	dfs = func(p int) {
		if len(vis) == k {
			tmp := make([]int, k)
			copy(tmp, vis)
			ans = append(ans, tmp)
			return
		}
		// 剪枝 ： 元素不足 k 个
		if len(vis)+n-p+1 < k {
			return
		}
		for i := p; i <= n; i++ {
			vis = append(vis, i)
			dfs(i + 1)
			vis = vis[:len(vis)-1]
		}
	}
	dfs(1)
	return ans
}

File: cpp/0077M-组合.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-26 13:39:15
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    vector<vector<int>> ans;
    vector<int> tmp;

    void dfs(int n, int k, int idx) {
        if (tmp.size() == k) {
            ans.push_back(tmp);
            return;
        }
        for (int i = idx; i <= n - (k - tmp.size()) + 1; i++) {
            tmp.push_back(i);
            dfs(n, k, i + 1);
            tmp.pop_back();
        }
    }

    vector<vector<int>> combine(int n, int k) {
        dfs(n, k, 1);
        return ans;
    }
};